// @ngneat/bind-query-params/esm2020/lib/BindQueryParamsFactory.mjs
import { Inject, Injectable } from '@angular/core';
import { BindQueryParamsManager } from './BindQueryParamsManager';
import { BIND_QUERY_PARAMS_OPTIONS } from './options';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class BindQueryParamsFactory {
    constructor(router, options) {
        this.router = router;
        this.options = options;
    }
    create(defs, createOptions) {
        return new BindQueryParamsManager(this.router, defs, this.options, createOptions);
    }
}
BindQueryParamsFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: BindQueryParamsFactory, deps: [{ token: i1.Router }, { token: BIND_QUERY_PARAMS_OPTIONS }], target: i0.ɵɵFactoryTarget.Injectable });
BindQueryParamsFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: BindQueryParamsFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: BindQueryParamsFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [BIND_QUERY_PARAMS_OPTIONS]
                }] }]; } });
